generator/src/generator/namespace.rs

```diff
@@ -4539,7 +4584,10 @@ impl<'ns, 'c> NamespaceGenerator<'ns, 'c> {
         // Deduced fields are not converter to their enum value
         if !was_deduced && self.use_enum_type_in_field(type_).is_some() {
             let rust_wire_type = self.type_to_rust_type(type_.type_.get_resolved());
-            format!("{}::from({}).serialize()", rust_wire_type, value)
+            format!(
+                "Option::<{}>::from({}).unwrap().serialize()",
```

Comment:
This is to handle cases where the enum has values larger than the wire type. I'm not sure if it is needed. There's https://github.com/psychon/x11rb/issues/543, but it is about a mask enum, which is not handled here.

Reply:
Hm, needed for example here:

`DeviceType` is serialized as `CARD8`:
https://github.com/psychon/x11rb/blob/a29b84d5c9d4db5dd1a08a7343764f6dc3b2c9c9/xcb-proto-1.14-1-g2b3559c/src/xinput.xml#L2379-L2386
`DeviceType` is serialized as `CARD16`:
https://github.com/psychon/x11rb/blob/a29b84d5c9d4db5dd1a08a7343764f6dc3b2c9c9/xcb-proto-1.14-1-g2b3559c/src/xinput.xml#L1752-L1767

So the `DeviceType` newtype wraps an `u16`, even though the defined values for the enum fits an `u8`.

Reply:
How is that possible? IIRC, we didn't handle that before and an enum could only be converted into `u8` if all its variants fit it into an `u8`.

Reply:
Note that the unwrap should be optimized away in cases where it is not needed.

Reply:
Well, `DeviceType` has values from 1 to 5. That fits into an `u8`, so the enum implements `From<Self> for u8`. But it also implements this conversion for larger types, i.e. `From<Self> for u16`. Thus, you can just do `u16::from(value)` instead of `Option<u16>::from(value).unwrap()`, or am I missing something?

Reply:
The issue here is that the new type does not wrap an `u8`, it wraps an `u16` (so there isn't `From<Self> for u8`), so it can infallibly parse the case where the wire type is `CARD16`. This is far from ideal, but X11 is hard. I'll try to think a better way of handling this...

Reply:
Uhm. Are you saying that the problem is turning an `CARD16` into an `u16`? Huh? Or do you mean `CARD8` as the wire type with `u16` as the smallest size that fits the enum? I'm confused and your example doesn't help me much here (since that enum does fit into `u8`).

Reply:
Indeed, the enum fits into an `u8`. However, there are some places where the wire type is actually larger (`u16`), so, our newtype uses an `u16` in order to be able to keep all values that come from the wire. Unfortunately, there are other places where the wire type is `u8`, so converting the newtype to `u8` is fallible.

Reply:
Ahhhh. Now I understand the problem. Sorry for being so blind. I totally missed that part of the problem.

Also: Urgh.

So, enums in xcb-proto do not have an inherent size / bit length. Perhaps we should go that way and make enums generic? I doubt that will result in nice code, but that's the first idea that came to my mind. (And the list of trait bounds will be huge... at least `T: From<u16>` expresses that something can store at least 16 bits.)

Just a quick first thought. All of this is a mess...

Reply:
I made some quick code to get which enums have this issue:
```
xproto BackingStore
xproto WindowClass
xproto Gravity
xproto StackMode
xproto AutoRepeatMode
render SubPixel
xinput DeviceType
```

Reply:
Most of `unwrap`s are now gone, the remaining only appear for the enums listed above.

Reply:
If it's just six special cases... how about adding them as special cases?

So... render's `SubPixel` appears only once in RENDER:
- in the reply to `QueryPictFormats` as `CARD32`
All other uses are in RandR:
- reply to `SetScreenConfig` as `CARD16`
- reply to `GetOutputInfo` as `CARD8`.
- event `ScreenChangeNotify` as `CARD16`
- struct `OutputChange` as `CARD8`

Okay...
- Your current version is to try to work around this. If I understood correctly, you currently make `SubPixel` a newtype around `u32`. This should actually work fine for `SubPixel` since this enum is only ever read from the wire.
- We could make only the six enums that are affected generic on the underlying type
- We could duplicate them and have `SubPixel`, `SubPixel_u16` and `SubPixel_u32` (I'm bad at naming things)
- [Below I came up with another idea: truncating the value when serialising]

The problem with your current version is when we have to serialise stuff. E.g. `DeviceType` appears in `struct`s as `u8` and `u16` and we have to "do something" there. Hm.

Another random idea would be: Just have a wrapping conversion? The newtype `DeviceType` is around an `u16` and does not implement `Serialize`. Instead, the code could just truncate the `u16` to an `u8`: `value as u8`. This would silently  discard the high bits, but so what? Garbage in, garbage out. Or something like that.

Oorrrr... `ConnectionError` has a `ParseError` variant. So, instead of panicing, sending the request could fail with `ConnectionError::ParseError(ParseError::ConversionFailed)`?

So much for my random ramblings. Perhaps one of these is a good idea. I don't know.
Also, I have a headache today. Sorry.

Reply:
I don't dislike the truncating conversion. It's the simplest to implement and produces the simplest API. The X11 protocol (and implementations) has been accumulating dust for too long and probably it's too late to build a perfect API on top of it.